File: javaDir/HandAlyzer.java
import java.util.Arrays;

public class HandAlyzer {

	enum HandRank {
		HIGH_CARD, PAIR, TWO_PAIR, THREE_OF_A_KIND, STRAIGHT,
		FLUSH, FULL_HOUSE, FOUR_OF_A_KIND, STRAIGHT_FLUSH, ROYAL_STRAIGHT_FLUSH;
	}

	//For importing all the hands
	Card[][] hands;
	//For splitting all hands into each players hand
	Card[] hand1 = new Card[5];
	Card[] hand2 = new Card[5];
	Card[] hand3 = new Card[5];
	Card[] hand4 = new Card[5];
	Card[] hand5 = new Card[5];
	Card[] hand6 = new Card[5];	

	//For tracking win and tie breaking info
	//The following are what each index is used for
	//[0] = HandRank //denoted as 0 (high card)) -> 9 (R straight flush)
	//[1] = Highest Rank card (or highest Rank of pair)
	//[2] = 2nd highest Rank card (if needed, else 0)
	//[3] = 3rd highest Rank card (if needed, else 0)
	//[4] = 4th highest Rank card (if needed, else 0)
	//[5] = 5th highest Rank card (if needed, else 0)
	//[6] = Suit of card in [1] // denoted as (0, D) (1, C) (2, H) (3, S)
	int[] scoreKeeperP1 = {0, 0, 0, 0, 0, 0, 0};
	int[] scoreKeeperP2 = {0, 0, 0, 0, 0, 0, 0};
	int[] scoreKeeperP3 = {0, 0, 0, 0, 0, 0, 0};
	int[] scoreKeeperP4 = {0, 0, 0, 0, 0, 0, 0};
	int[] scoreKeeperP5 = {0, 0, 0, 0, 0, 0, 0};
	int[] scoreKeeperP6 = {0, 0, 0, 0, 0, 0, 0};
	int[][] scoreKeeperAll =  {scoreKeeperP1, scoreKeeperP2, scoreKeeperP3,
			scoreKeeperP4, scoreKeeperP5, scoreKeeperP6};


	//Holds HandRank enums in player order 1-6
	HandRank[] playerHandRanks = new HandRank[6];



	//Constructor ----------------------------------------//
	//Can only be constructed with [6 player hands][5 Cards a hand]
	public HandAlyzer(Card[][] hands) {
		this.hands = hands;
		for (int i = 0; i < 5; i++) {
			hand1[i] = hands[0][i];
			hand2[i] = hands[1][i];
			hand3[i] = hands[2][i];
			hand4[i] = hands[3][i];
			hand5[i] = hands[4][i];
			hand6[i] = hands[5][i];
		}
	}

	//Start of Methods------------------------------------------------//


	//------------------------------------------------//
	//Converts HandRank enum to String
	public String handRanktoString(HandRank rank) {
		String rankString = "";
		switch (rank) {
		case HIGH_CARD: rankString += "High Card"; break;
		case PAIR: rankString += "Pair"; break;
		case TWO_PAIR: rankString += "Two Pair"; break;
		case THREE_OF_A_KIND: rankString += "Three of a Kind"; break;
		case STRAIGHT: rankString += "Straight"; break;
		case FLUSH: rankString += "Flush"; break;
		case FULL_HOUSE: rankString += "Full House"; break;
		case FOUR_OF_A_KIND: rankString += "Four of a Kind"; break;
		case STRAIGHT_FLUSH: rankString += "Straight Flush"; break;
		case ROYAL_STRAIGHT_FLUSH: rankString += "Royal Straight Flush"; break;
		}
		return rankString;
	}

	//------------------------------------------------//
	public void getFinalScorePrint() {
		String[] handStrings = new String[6];

		//Get hand rank and build strings for printing
		for (int i = 0; i < 6; i++) {
			int[] scoreKeeper = new int[7];
			HandRank rank = getHandRank(hands[i], scoreKeeper);
			playerHandRanks[i] = rank;
			this.scoreKeeperAll[i][0] = rank.ordinal();

			for (int j = 1; j < 7; j++){
				this.scoreKeeperAll[i][j] = scoreKeeper[j]; 
			}

			//Build out Hands to print, stored in order of players 1->6
			StringBuilder handString = new StringBuilder();
			for (Card card : hands[i]) {
				handString.append(String.format("%4s", card.toString()));
			}
			handStrings[i] = handString.toString();
		}

		// //DEBUGGING PURPOSES
		// //Adds [0]1[2][3][4][5][6] section to printout, with the values of the player scoreKeeperAll[player][0-6]
		// for (int i = 0; i < 6 ; i++) {
		// 	String temp = handStrings[i] + " ";
		// 	for (int j = 0; j < 7; j++) {
		// 		if (this.scoreKeeperAll[i][j] <10 ) {
		// 			temp += "[0" + this.scoreKeeperAll[i][j] + "]";
		// 		} else temp += "[" + this.scoreKeeperAll[i][j] + "]";
		// 	}
		// 	handStrings[i] = temp;
		// }
		// //END DEBUGGING STRING APPENDER

		//////////////Sort out winners and implement tie breaking
		for (int playerRankIndexLimit = 0; playerRankIndexLimit < 6 ;  playerRankIndexLimit++){
			for(int currentPlayer = 0; currentPlayer < 5; currentPlayer++){
				if (this.scoreKeeperAll[currentPlayer][0] < this.scoreKeeperAll[currentPlayer + 1][0]){
					swapHands(currentPlayer, currentPlayer + 1, handStrings);
				} 
				//Tie Breaking
				else if (this.scoreKeeperAll[currentPlayer][0] == this.scoreKeeperAll[currentPlayer + 1][0]){
					for (int criteriaIndex = 1; criteriaIndex < 7; criteriaIndex++){
						if(this.scoreKeeperAll[currentPlayer][criteriaIndex] < this.scoreKeeperAll[currentPlayer + 1][criteriaIndex]){
							swapHands(currentPlayer, currentPlayer + 1, handStrings);
							break;

						} else if (this.scoreKeeperAll[currentPlayer][criteriaIndex] > this.scoreKeeperAll[currentPlayer + 1][criteriaIndex]){
							break;
						}
					}
				}
			}
		}

		//Print the sorted hands with their ranks
		System.out.println("--- WINNING HAND ORDER ---");
		for (int i = 0; i < 6; i++) {
			System.out.println(handStrings[i] + " - " + handRanktoString(playerHandRanks[i])
			//Debugging step - print out playerHandRanks[]

					);
		}

	}
	//-------------------
	// Helper method to swap hands and corresponding data
	private void swapHands(int player1, int player2, String[] handStrings) {
		int[] tempScoreKeeper = scoreKeeperAll[player1];
		scoreKeeperAll[player1] = scoreKeeperAll[player2];
		scoreKeeperAll[player2] = tempScoreKeeper;

		HandRank tempRank = playerHandRanks[player1];
		playerHandRanks[player1] = playerHandRanks[player2];
		playerHandRanks[player2] = tempRank;

		String tempHand = handStrings[player1];
		handStrings[player1] = handStrings[player2];
		handStrings[player2] = tempHand;
	}
	//------------------------------------------------//
	//Checks one hand of 5 cards for its HandRank
	//Currently, assigns High Card if no other HandRank fits,
	//but it does not rank at all
	public HandRank getHandRank(Card[] hand, int[] scoreKeeper) {
		if (isStraightRoyalFlush(hand, scoreKeeper)) {
			return HandRank.ROYAL_STRAIGHT_FLUSH;
		} else if (isStraightFlush(hand, scoreKeeper)) {
			return HandRank.STRAIGHT_FLUSH;
		} else if (isFourOfAKind(hand, scoreKeeper)) {
			return HandRank.FOUR_OF_A_KIND;
		} else if (isFullHouse(hand, scoreKeeper)) {
			return HandRank.FULL_HOUSE;
		} else if (isFlush(hand, scoreKeeper)) {
			return HandRank.FLUSH;
		} else if (isStraight(hand, scoreKeeper)) {
			return HandRank.STRAIGHT;
		} else if (isThreeOfAKind(hand, scoreKeeper)) {
			return HandRank.THREE_OF_A_KIND;
		} else if (numOfPairs(hand, scoreKeeper) == 2) {
			return HandRank.TWO_PAIR;
		} else if (numOfPairs(hand, scoreKeeper) == 1) {
			return HandRank.PAIR;
		} else if (scoreKeeper[0] == 0) {
			getHighCard(hand, scoreKeeper);
			getHighCardTieBreakers(hand, scoreKeeper);

		}		
		return HandRank.HIGH_CARD;
	}

	//------------------------------------------------//
	//Hand Rank Identifiers---------------------------//
	//------------------------------------------------//

	//Checks for the highest value card in a hand
	//First, by rank A,2,3,...Q,K (L -> H)
	//Then if a tie, by suit D,C,H,S (L -> H)
	public Card getHighCard(Card[] playerHand, int[] scoreKeeper) {
		Card highCard = new Card();
		boolean isAce = false;
		for (Card card : playerHand) {
			//check for aces, if found only looks for other possible aces to compare
			if (card.rank.ordinal() == 0){
				if (highCard.rank.ordinal() != 0){
					highCard = card;
				}
				if (card.suit.ordinal() > highCard.suit.ordinal()){
					highCard = card;
				}
				isAce = true;
			}
			//else, find highest rank card
			else if (highCard.rank.ordinal() < card.rank.ordinal())
			{
				highCard = card;
			}
			//if tied for rank, compare suit
			else if ((highCard.rank.ordinal() == card.rank.ordinal()) 
					&& (highCard.suit.ordinal() < card.suit.ordinal()))
			{
				highCard = card;
			}

<<<<<<< HEAD

=======
>>>>>>> 0449f2369367c623e90b0a24aebb28eed3801559
		}	

		//Keeps Aces High for high card
		if (isAce){
			scoreKeeper[1] = 14;
		} else scoreKeeper[1] = highCard.rank.ordinal();

		scoreKeeper[6] = highCard.suit.ordinal();

		return highCard;
	}

	public void getHighCardTieBreakers(Card[] hand, int[] scoreKeeper) {
		int[] ranks = new int[5];
		for (int i = 0; i < 5; i++) {
			ranks[i] = hand[i].rank.ordinal();
			//For ties, ranks ace as highest value
			if (ranks[i] == 0){
<<<<<<< HEAD
				ranks[i] = 13;
=======
				ranks[i] = 14;
>>>>>>> 0449f2369367c623e90b0a24aebb28eed3801559
			}
		}
		Arrays.sort(ranks); //sorted [low to high]
		int sKIndex = 2;

<<<<<<< HEAD
		//Reverses the array from [low to high], to, [high to low]
		for (int i = 4; i > 0; i--) {
=======
		//Reverses the array from [low to high], to, [high to low] when assigning values to scoreKeeper[]
		for (int i = 3; i >= 0; i--) {
>>>>>>> 0449f2369367c623e90b0a24aebb28eed3801559
			scoreKeeper[sKIndex] = ranks[i];
			sKIndex++;
		}

	}

	//------------------------------------------------//
	public boolean isStraight(Card[] hand, int[] scoreKeeper) {
		int[] ranks = new int[5];
		Card highestCard = new Card();
		for (int i = 0; i < 5; i++) {
			ranks[i] = hand[i].rank.ordinal();

			if (hand[i].rank.ordinal() > highestCard.rank.ordinal()) {
				highestCard = hand[i];
			}
		}

		Arrays.sort(ranks);

		if (isStraightRoyal(hand, scoreKeeper)) {
			return true;
		} else {
			for (int i = 0; i < 4; i++) {
				if (ranks[i] + 1 != ranks[i + 1]) {
					return false;
				}
			}
		}

		getHighCard(hand, scoreKeeper);
		return true;
	}

	//------------------------------------------------//
	public boolean isFlush(Card[] hand, int[] scoreKeeper) {
		for (int i = 0; i < 4; i++) {
			if (hand[i].suit.ordinal() != hand[i + 1].suit.ordinal()) {
				return false;
			}
		}
		getHighCard(hand, scoreKeeper);
		return true;
	}

	//------------------------------------------------//
	//This method does NOT account for wrap around flushes!
	public boolean isStraightFlush(Card[] hand, int[] scoreKeeper) {
		return (isFlush(hand, scoreKeeper) && isStraight(hand, scoreKeeper));                                                                                                                                                                                
	}

	//------------------------------------------------//
	public boolean isStraightRoyalFlush(Card[] hand, int[] scoreKeeper) {
		int[] countsPerRank = new int[13];
		int[] suitPerCard = new int[5];
		int loopCounter = 0;
		int[] cardsToCheck = {Rank.A.ordinal(), 
				Rank.TEN.ordinal(), 
				Rank.J.ordinal(),
				Rank.Q.ordinal(),
				Rank.K.ordinal()};

		for (Card card : hand) {
			countsPerRank[card.rank.ordinal()]++;
			suitPerCard[loopCounter] = card.suit.ordinal();
			loopCounter++;
		}
		for (int i : cardsToCheck) {
			if (countsPerRank[i] != 1) {
				return false;
			}
		}

		for (int i = 0; i < 4; i++) {
			if (suitPerCard[i] != suitPerCard[i + 1]) {
				return false;
			}
		}
		
		scoreKeeper[6] = hand[0].suit.ordinal(); 		
		return (true);
	}

	//------------------------------------------------//
	//To solve edge case where (non-flush, royal) Straights were not identified
	//Catches straights in order of "10, J, Q, K, A" not caught by isStraight
	public boolean isStraightRoyal(Card[] hand, int[] scoreKeeper) {
		int[] countsPerRank = new int[13];
		int[] cardsToCheck = {Rank.A.ordinal(), 
				Rank.TEN.ordinal(), 
				Rank.J.ordinal(),
				Rank.Q.ordinal(),
				Rank.K.ordinal()};
		int aceSuit = 0;

		for (Card card : hand) {
			countsPerRank[card.rank.ordinal()]++;
			if (card.rank.ordinal() == 0) aceSuit = card.suit.ordinal();				
		}

		for (int i : cardsToCheck) {
			if (countsPerRank[i] != 1) {
				return false;
			}
		}

		scoreKeeper[1] = 14; //Ace High value hard coded, since it will be in every SRF
		scoreKeeper[6] = aceSuit; 		
		return (true);
	}
	//------------------------------------------------//
	//Returns int for num of pairs, unlike rest returning bool!
	public int numOfPairs(Card[] hand, int[] scoreKeeper) {
		int numofPairs = 0;
		int iterator = 0;
		int kickerRank = 0;
		int[] rankofPair = new int[2];
		int[] countsPerRank = new int[13];
		for (Card card : hand) {
			countsPerRank[card.rank.ordinal()]++;
		}

		for (int i = 0; i < 13; i++) {
			if (countsPerRank[i] == 2) {
				numofPairs++;
				//check for aces, make high
				if (i == 0){
					rankofPair[iterator] = 14;
				} else rankofPair[iterator] = i;
			
				iterator++;
			}
			else if (countsPerRank[i] == 1) {
				kickerRank = i;
			}
		}

		for (Card card : hand) {
			if (card.rank.ordinal() == kickerRank) {
				scoreKeeper[6] = card.suit.ordinal();
			}
		}

		if (numofPairs == 1) {
			//UPDATE SCOREKEEPER
			int[] ranks = new int[5];
			for (int i = 0; i < 5; i++) {
				ranks[i] = hand[i].rank.ordinal();
				//For ties, ranks ace as highest value
				if (ranks[i] == 0){
					ranks[i] = 14;
				}
			}
			Arrays.sort(ranks); //sorted [low to high]
			int sKIndex = 2;

			//Reverses the array from [low to high], to, [high to low] when assigning values to scoreKeeper[]
			//Only assigns value if the rank does not match the pair
			for (int i = 4; i >= 0; i--) {
				if (ranks[i] != rankofPair[0]){
				scoreKeeper[sKIndex] = ranks[i];
				sKIndex++;
				}
			}
			
			//FINISHED UPDATE
			scoreKeeper[1] = rankofPair[0];

		} else if (numofPairs == 2) {
			scoreKeeper[1] = rankofPair[1]; //higher value pair first
			scoreKeeper[2] = rankofPair[0]; //lower value pair second
		}

		return numofPairs;
	}

	//------------------------------------------------//
	public boolean isThreeOfAKind(Card[] hand, int[] scoreKeeper) {
		int[] countsPerRank = new int[13];
		for (Card card : hand) {
			countsPerRank[card.rank.ordinal()]++;
		}

		for (int i = 0; i < 13; i++) {
			if (countsPerRank[i] == 3) {
				getHighCardTieBreakers(hand, scoreKeeper);
				getHighCard(hand, scoreKeeper);
				return true;
			}
		}
		return false;
	}

	//------------------------------------------------//
	public boolean isFullHouse(Card[] hand, int[] scoreKeeper) {
		if (isThreeOfAKind(hand, scoreKeeper) && (numOfPairs(hand, scoreKeeper) == 1)) {
			getHighCard(hand, scoreKeeper);
			return true; 
		}
		return false;
	}

	//------------------------------------------------//
	public boolean isFourOfAKind(Card[] hand, int[] scoreKeeper) {
		int[] countsPerRank = new int[13];
		for (Card card : hand) {
			countsPerRank[card.rank.ordinal()]++;
		}

		for (int i = 0; i < 13; i++) {
			if (countsPerRank[i] == 4) {
				getHighCard(hand, scoreKeeper);
				return true;
			}
		}
		return false;
	}
	//------------------------------------------------//

}

File: pyDir/HandAlyzer.py
# Willem Ethan Wilcox, CSC 330, Dr. Pounds, Mercer University
from Deck import Card, Deck


class HandAlyzer:
    def __init__(self, hands):
        # creates 2 2d arrays
        # hands[6][5] imported from tableTest
        self.hands = hands
        # scoreKeeper[6][7]
        # scoreKeeper[Player][0-6] where [0-6] is tiebreaking criteria
        self.score_keeper_all = [[0] * 7 for _ in range(6)]

        # tracks the strings for each player, starts players 1->6
        self.player_hand_ranks = [0] * 6
        self.hand_rank_strings = [
            "High Card",  # index 0
            "Pair",
            "Two Pair",
            "Three of a Kind",
            "Straight",
            "Flush",
            "Full House",
            "Four of a Kind",
            "Straight Flush",
            "Royal Straight Flush"  # index 9
        ]

    def get_final_score_print(self):
        hand_strings = [""] * 6

        # get hand ranks and build strings for printing
        for i in range(0, 6, 1):
            score_keeper = [0] * 7
            current_hand_rank = self.get_hand_rank(self.hands[i], score_keeper)
            self.player_hand_ranks[i] = current_hand_rank
            self.score_keeper_all[i] = score_keeper
            self.score_keeper_all[i][0] = current_hand_rank

            # Build out hands to print, stored in order of players 1->6
            hand_string = ""
            for card in self.hands[i]:
                if card.get_rank_int() == 9: # hard coded adjustment for printing rank 10 cards
                    hand_string += f"{str(card):<4}"
                else:
                    hand_string += " "
                    hand_string += f"{str(card):<3}"
            hand_strings[i] = hand_string

        # Sort out winners and implement tie-breaking
        for player_rank_index_limit in range(0, 6, 1):
            for current_player in range(0, 5, 1):
                if self.score_keeper_all[current_player][0] < self.score_keeper_all[current_player + 1][0]:
                    self.swap_hands(current_player, current_player + 1, hand_strings)
                # Tie Breaking
                elif self.score_keeper_all[current_player][0] == self.score_keeper_all[current_player + 1][0]:
                    for criteria_index in range(1, 7, 1):
                        if self.score_keeper_all[current_player][criteria_index] < \
                                self.score_keeper_all[current_player + 1][criteria_index]:
                            self.swap_hands(current_player, current_player + 1, hand_strings)
                            break
                        elif self.score_keeper_all[current_player][criteria_index] > \
                                self.score_keeper_all[current_player + 1][criteria_index]:
                            break

        # Print the sorted hands with their ranks
        print("--- WINNING HAND ORDER ---")
        for i in range(0, 6, 1):
            print(f"{hand_strings[i]} - {self.hand_rank_strings[self.score_keeper_all[i][0]]}")

    # Helper Method for swapping corresponding player data in get_final_score_print
    def swap_hands(self, player1, player2, hand_strings):
        temp_score_keeper = self.score_keeper_all[player1]
        self.score_keeper_all[player1] = self.score_keeper_all[player2]
        self.score_keeper_all[player2] = temp_score_keeper

        temp_rank = self.player_hand_ranks[player1]
        self.player_hand_ranks[player1] = self.player_hand_ranks[player2]
        self.player_hand_ranks[player2] = temp_rank

        temp_hand = hand_strings[player1]
        hand_strings[player1] = hand_strings[player2]
        hand_strings[player2] = temp_hand

    def get_hand_rank(self, hand, score_keeper):
        if self.is_straight_royal_flush(hand, score_keeper):
            return 9
        elif self.is_straight_flush(hand, score_keeper):
            return 8
        elif self.is_four_of_a_kind(hand, score_keeper):
            return 7
        elif self.is_full_house(hand, score_keeper):
            return 6
        elif self.is_flush(hand, score_keeper):
            return 5
        elif self.is_straight(hand, score_keeper):
            return 4
        elif self.is_three_of_a_kind(hand, score_keeper):
            return 3
        elif self.num_of_pairs(hand, score_keeper) == 2:
            return 2
        elif self.num_of_pairs(hand, score_keeper) == 1:
            return 1
        else:
            self.get_high_card(hand, score_keeper)
            self.get_high_card_tie_breakers(hand, score_keeper)
            return 0

    # ------------------------------------------------ Hand Rank Identifier Methods
    def get_high_card(self, hand, score_keeper):
        # Checks for the highest value card in a hand
        # First, by rank A, 2, 3, ... Q, K (L -> H)
        # Then if a tie, by suit D, C, H, S (L -> H)

        # initialize high_card as lowest value card
        high_card = Card("A", "D")
        is_ace = False

        for card in hand:
            # checks for ace
            if card.get_rank_int() == 0:
                if high_card.get_rank_int() != 0:
                    high_card = card
                elif high_card.get_suit_int() < card.get_suit_int():
                    high_card = card
                is_ace = True
            # check for everything else
            elif high_card.get_rank_int() < card.get_rank_int():
                high_card = card
            elif high_card.get_suit_int() < card.get_suit_int() and high_card.get_rank_int() == card.get_rank_int():
                high_card = card

        # update score_keeper, check for aces and hard code to 13
        if is_ace:
            score_keeper[1] = 13
        else:
            score_keeper[1] = high_card.get_rank_int()
        score_keeper[6] = high_card.get_suit_int()

        return high_card

    def get_high_card_tie_breakers(self, hand, score_keeper):
        current_hand_ranks = [0]*5
        iterator = 0

        # get values of ranks in hand, store to sort
        for card in hand:
            current_hand_ranks[iterator] = card.get_rank_int()
            if current_hand_ranks[iterator] == 0:
                current_hand_ranks[iterator] = 13

        current_hand_ranks.sort(reverse=True)
        iterator = 2
        for i in range(0, 4, 1):
            score_keeper[iterator] = current_hand_ranks[i]
            iterator += 1

    def is_straight(self, hand, score_keeper):
        current_hand_ranks = [card.get_rank_int() for card in hand]
        current_hand_ranks.sort()

        for i in range(0, 4, 1):
            if current_hand_ranks[i] + 1 != current_hand_ranks[i + 1]:
                return False

        self.get_high_card(hand, score_keeper)

    def is_flush(self, hand, score_keeper):
        current_hand_suits = [card.get_suit_int() for card in hand]

        for i in range(0, 4, 1):
            if current_hand_suits[i] != current_hand_suits[i + 1]:
                return False

        self.get_high_card(hand, score_keeper)
        return True

    def is_straight_flush(self, hand, score_keeper):
        return self.is_flush(hand, score_keeper) and self.is_straight(hand, score_keeper)

    def is_straight_royal_flush(self, hand, score_keeper):
        # Makes array showings how many of each rank are in hand. A=0, 2=1, ... K=12
        cards_in_hand = [0] * 13
        current_hand_ranks = [card.get_rank_int() for card in hand]
        for i in current_hand_ranks:
            cards_in_hand[i] += 1

        # defined array of indices for SRF
        check_for_rsf = [0, 9, 10, 11, 12]

        for i in check_for_rsf:
            if cards_in_hand[i] != 1:
                return False

        if self.is_flush(hand, score_keeper):
            score_keeper[2] = 13  # hard coded for ace
            score_keeper[6] = hand[0].get_suit_int()
            return True

        return False

    def num_of_pairs(self, hand, score_keeper):
        number_of_pairs = 0
        rank_of_pair = [0]*2
        kicker_rank = 0

        # Makes array showings how many of each rank are in hand. A=0, 2=1, ... K=12
        cards_in_hand = [0]*13
        current_hand_ranks = [card.get_rank_int() for card in hand]
        for i in current_hand_ranks:
            cards_in_hand[i] += 1

        for i in range(0, 13, 1):
            if cards_in_hand[i] == 2:
                rank_of_pair[number_of_pairs] = i
                number_of_pairs += 1
            if cards_in_hand[i] == 1:
                kicker_rank = i

        if number_of_pairs == 1:
            current_hand_ranks.sort(reverse=True)
            iterator = 2
            score_keeper[1] = rank_of_pair[0]
            score_keeper[6] = kicker_rank
            for i in range(0, 5, 1):
                if current_hand_ranks[i] != rank_of_pair[0]:
                    score_keeper[iterator] = current_hand_ranks[i]
        elif number_of_pairs == 2:
            score_keeper[1] = rank_of_pair[1]
            score_keeper[2] = rank_of_pair[0]
            score_keeper[6] = kicker_rank

        return number_of_pairs

    def is_three_of_a_kind(self, hand, score_keeper):
        # Makes array showings how many of each rank are in hand. A=0, 2=1, ... K=12
        cards_in_hand = [0]*13
        current_hand_ranks = [card.get_rank_int() for card in hand]
        for i in current_hand_ranks:
            cards_in_hand[i] += 1

        for i in range(0, 13, 1):
            if cards_in_hand[i] == 3:
                self.get_high_card_tie_breakers(hand, score_keeper)
                self.get_high_card(hand, score_keeper)
                return True

        return False

    def is_full_house(self, hand, score_keeper):
        if self.is_three_of_a_kind(hand, score_keeper) and self.num_of_pairs(hand, score_keeper) == 1:
            self.get_high_card(hand, score_keeper)
            return True
        else:
            return False

    def is_four_of_a_kind(self, hand, score_keeper):
        # Makes array showings how many of each rank are in hand. A=0, 2=1, ... K=12
        cards_in_hand = [0]*13
        current_hand_ranks = [card.get_rank_int() for card in hand]
        for i in current_hand_ranks:
            cards_in_hand[i] += 1

        for i in range(0, 13, 1):
            if cards_in_hand[i] == 4:
                self.get_high_card_tie_breakers(hand, score_keeper)
                self.get_high_card(hand, score_keeper)
                return True

        return False
